codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/shapevisitor/SmithyToDafnyShapeVisitor.java

```diff
@@ -465,7 +466,94 @@ public String doubleShape(DoubleShape shape) {
 
     @Override
     public String unionShape(UnionShape shape) {
-        return "Unionnnnnn";
+        String internalDafnyType = DafnyNameResolver.getDafnyType(shape, context.symbolProvider().toSymbol(shape));
+        writer.addImportFromModule("github.com/dafny-lang/DafnyRuntimeGo", "dafny");
+        String returnString = """
+            func() Wrappers.Option {
+                switch %s.(type) {""".formatted(dataSource);
+        for (var member : shape.getAllMembers().values()) {
+            String memberName = context.symbolProvider().toMemberName(member);
+            String createMemberFunction = memberName.replace(shape.getId().getName() + "Member", "Create_")+"_";
+            Shape targetShape = context.model().expectShape(member.getTarget());
+            String someWrapIfRequired = "Wrappers.Companion_Option_.Create_Some_(companion.%s(%s))";
+            returnString += """
+                    case *%s.%s:
+                        var companion = %s
+                        var inputToConversion = %s
+                        if inputToConversion.UnwrapOr(nil) == nil { return Wrappers.Companion_Option_.Create_None_() }
+                    """.formatted(
+                            SmithyNameResolver.smithyTypesNamespace(shape),
+                            context.symbolProvider().toMemberName(member),
+                            internalDafnyType.replace(shape.getId().getName(), "CompanionStruct_" + shape.getId().getName() + "_{}"),
+                            targetShape.accept(
+                                    new SmithyToDafnyShapeVisitor(
+                                        context,dataSource + ".(*" + SmithyNameResolver.smithyTypesNamespace(shape) + "." + context.symbolProvider().toMemberName(member) + ").Value", writer, isConfigShape, true, false
+                                    )
+                                )
+                        );
+            
+            if(targetShape.isMapShape()){
```

Comment:
Separating it looks more readable we can decrease this to a single block of code but I don't feel it will be more readable if done so but I might be biased as I wrote this code